File: EcoDietMaker/interface/prompts.py
"""Interactive prompts used by the CLI.

Order enforced by `collect_user_constraints`:
1) Current calories → 2) Max calories → 3) Satisfied cravings → 4) Current craving.
"""

from constants import (
    TASTINESS_MULTIPLIERS,
    TASTINESS_NAMES,
)


def prompt_for_cravings_satisfied():
    """Prompt the user for how many cravings have been satisfied today.

    Returns
    -------
    int
        Number of cravings satisfied (0+).
    """
    while True:
        value = input("How many cravings have you satisfied today? (0+) > ").strip()
        if value == "":
            return 0
        try:
            num = int(value)
            if num >= 0:
                return num
            print("Please enter 0 or a positive number.")
        except ValueError:
            print("Invalid input. Enter a whole number.")


def prompt_current_calories() -> int:
    """Ask for calories already consumed (>= 0)."""
    while True:
        try:
            val = int(input("How many calories have you already consumed? > ").strip())
            if val < 0:
                print("Calories cannot be negative.")
            else:
                return val
        except ValueError:
            print("That doesn't seem to be a number.")


def prompt_max_calories(
    current_cal: int,
) -> int:
    """Ask for max calorie limit (>= current)."""
    while True:
        try:
            val = int(input("What is your maximum calorie limit? > ").strip())
            if val < current_cal:
                print("Maximum must be greater than or equal to current intake.")
            else:
                return val
        except ValueError:
            print("That doesn't seem to be a number.")


def collect_user_constraints():
    """Collect user-defined constraints for meal planning.

    Order:
    1) Current calories
    2) Max calories
    3) Satisfied cravings
    4) Current craving
    (Rating unknown foods is handled during load.)

    Returns
    -------
    tuple[list[str], int, int]
        ``(cravings, cravings_satisfied, remaining_calories)``.
    """
    current_cal = prompt_current_calories()
    max_cal = prompt_max_calories(current_cal)
    cravings_satisfied = prompt_for_cravings_satisfied()
    print("Enter your current craving (press Enter to skip):")
    entry = input("> ").strip()
    cravings = [entry.lower()] if entry else []
    remaining_calories = max_cal - current_cal
    return (cravings, cravings_satisfied, remaining_calories)


def prompt_for_tastiness(
    food_name,
):
    """Prompt user for tastiness rating for a given food item.

    Parameters
    ----------
    food_name : str
        Name of the food item to rate.

    Returns
    -------
    int
        Tastiness rating in ``{-3, -2, -1, 0, 1, 2, 3}``, or ``99`` if skipped/unknown.
    """
    print(f"[PROMPT] Enter tastiness rating for '{food_name}' (-3 to 3, or 99 for unknown):")

    # Show human labels for valid ratings; exclude 99 (unknown) from the hint line
    print("  Hints:", ", ".join(f"{key}: {TASTINESS_NAMES[key]}" for key in sorted(TASTINESS_NAMES) if key != 99))

    # Accept blank to keep tastiness as unknown (99); otherwise loop until a valid int in range
    while True:
        value = input("> ").strip()
        if value == "":
            return 99  # Skip and keep as unknown
        try:
            value = int(value)

            # Validate against constants, not hardcoded bounds, so changing the scale only touches constants.py
            if value in TASTINESS_MULTIPLIERS:
                return value
            print("Invalid value. Valid: -3 to 3, or 99 for unknown.")
        except ValueError:
            print("Enter an integer or press Enter to skip.")


def prompt_yes_no(
    prompt: str,
    default: bool = True,
) -> bool:
    """Prompt the user for a yes/no response.

    Parameters
    ----------
    prompt : str
        The question to display.
    default : bool, optional
        Default value if the user presses Enter, by default ``True``.

    Returns
    -------
    bool
        ``True`` for yes, ``False`` for no.
    """

    # Conventional Y/n or y/N rendering with the default capitalized
    suffix = " (Y/n) > " if default else " (y/N) > "
    while True:
        # Normalize input (strip + lowercase); blank means "take the default"
        resp = input(prompt + suffix).strip().lower()
        if resp == "":
            return default
        if resp in ("y", "yes"):
            return True
        if resp in ("n", "no"):
            return False
        print("Please enter yes or no (y/n).")

File: src/interface/prompts.rs
use dialoguer::{Confirm, Input, Select};
use strsim::jaro_winkler;

use crate::error::{EcoError, Result};
use crate::models::Food;
use crate::planner::constants::tastiness_name;

/// Prompt for current calories already consumed.
pub fn prompt_current_calories() -> Result<f64> {
    let input: String = Input::new()
        .with_prompt("How many calories have you already consumed today?")
        .default("0".to_string())
        .interact_text()?;

    input
        .parse()
        .map_err(|_| EcoError::InvalidInput("Invalid number".to_string()))
}

/// Prompt for maximum calorie limit.
pub fn prompt_max_calories(current: f64) -> Result<f64> {
    let input: String = Input::new()
        .with_prompt("What is your maximum calorie limit for today?")
        .default("2500".to_string())
        .interact_text()?;

    let max: f64 = input
        .parse()
        .map_err(|_| EcoError::InvalidInput("Invalid number".to_string()))?;

    if max < current {
        return Err(EcoError::InvalidInput(
            "Max calories must be >= current calories".to_string(),
        ));
    }

    Ok(max)
}

/// Prompt for number of cravings already satisfied today.
pub fn prompt_cravings_satisfied() -> Result<u32> {
    let input: String = Input::new()
        .with_prompt("How many cravings have you satisfied today?")
        .default("0".to_string())
        .interact_text()?;

    input
        .parse()
        .map_err(|_| EcoError::InvalidInput("Invalid number".to_string()))
}

/// Prompt for current cravings with fuzzy matching.
pub fn prompt_cravings(available_foods: &[&Food]) -> Result<Vec<String>> {
    let mut cravings = Vec::new();

    loop {
        let input: String = Input::new()
            .with_prompt("Enter a craving (or press Enter to finish)")
            .allow_empty(true)
            .interact_text()?;

        let input = input.trim();
        if input.is_empty() {
            break;
        }

        // Try exact match first (case-insensitive)
        let exact_match = available_foods
            .iter()
            .find(|f| f.name.to_lowercase() == input.to_lowercase());

        if let Some(food) = exact_match {
            cravings.push(food.name.clone());
            println!("Added: {}", food.name);
            continue;
        }

        // Try fuzzy matching
        let mut candidates: Vec<(&Food, f64)> = available_foods
            .iter()
            .map(|f| (*f, jaro_winkler(&f.name.to_lowercase(), &input.to_lowercase())))
            .filter(|(_, score)| *score > 0.7)
            .collect();

        candidates.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        if candidates.is_empty() {
            println!("No matching food found for '{}'", input);
            continue;
        }

        if candidates.len() == 1 {
            let food = candidates[0].0;
            let confirm = Confirm::new()
                .with_prompt(format!("Did you mean '{}'?", food.name))
                .default(true)
                .interact()?;

            if confirm {
                cravings.push(food.name.clone());
                println!("Added: {}", food.name);
            }
        } else {
            // Multiple matches - let user select
            let options: Vec<String> = candidates
                .iter()
                .take(5)
                .map(|(f, _)| f.name.clone())
                .collect();

            let mut selection_options = options.clone();
            selection_options.push("None of these".to_string());

            let selection = Select::new()
                .with_prompt("Which did you mean?")
                .items(&selection_options)
                .default(0)
                .interact()?;

            if selection < options.len() {
                cravings.push(options[selection].clone());
                println!("Added: {}", options[selection]);
            }
        }
    }

    Ok(cravings)
}

/// Prompt for tastiness rating.
pub fn prompt_tastiness(food_name: &str) -> Result<i8> {
    let options = vec![
        format!("-3 ({})", tastiness_name(-3)),
        format!("-2 ({})", tastiness_name(-2)),
        format!("-1 ({})", tastiness_name(-1)),
        format!(" 0 ({})", tastiness_name(0)),
        format!(" 1 ({})", tastiness_name(1)),
        format!(" 2 ({})", tastiness_name(2)),
        format!(" 3 ({})", tastiness_name(3)),
        format!("99 ({})", tastiness_name(99)),
    ];

    let selection = Select::new()
        .with_prompt(format!("Rate the tastiness of '{}'", food_name))
        .items(&options)
        .default(3) // neutral (0)
        .interact()?;

    Ok(match selection {
        0 => -3,
        1 => -2,
        2 => -1,
        3 => 0,
        4 => 1,
        5 => 2,
        6 => 3,
        7 => 99,
        _ => 0,
    })
}

/// Prompt for yes/no confirmation.
pub fn prompt_yes_no(prompt: &str, default: bool) -> Result<bool> {
    Ok(Confirm::new()
        .with_prompt(prompt)
        .default(default)
        .interact()?)
}

/// Collect all user constraints for meal planning.
pub fn collect_user_constraints(
    available_foods: &[&Food],
) -> Result<(Vec<String>, u32, f64)> {
    let current_cal = prompt_current_calories()?;
    let max_cal = prompt_max_calories(current_cal)?;
    let cravings_satisfied = prompt_cravings_satisfied()?;
    let cravings = prompt_cravings(available_foods)?;

    let remaining = max_cal - current_cal;

    Ok((cravings, cravings_satisfied, remaining))
}
